src/datafusion/helpers.rs

```diff
@@ -0,0 +1,228 @@
+            }
+        }
+        Expr::Literal(_)
+        | Expr::Alias(_)
+        | Expr::OuterReferenceColumn(_, _)
+        | Expr::ScalarVariable(_, _)
+        | Expr::Not(_)
+        | Expr::IsNotNull(_)
+        | Expr::IsNull(_)
+        | Expr::IsTrue(_)
+        | Expr::IsFalse(_)
+        | Expr::IsUnknown(_)
+        | Expr::IsNotTrue(_)
+        | Expr::IsNotFalse(_)
+        | Expr::IsNotUnknown(_)
+        | Expr::Negative(_)
+        | Expr::Cast { .. }
+        | Expr::TryCast { .. }
+        | Expr::BinaryExpr { .. }
+        | Expr::Between { .. }
+        | Expr::Like { .. }
+        | Expr::SimilarTo { .. }
+        | Expr::InList { .. }
+        | Expr::Exists { .. }
+        | Expr::InSubquery(_)
+        | Expr::ScalarSubquery(_)
+        | Expr::GetIndexedField { .. }
+        | Expr::GroupingSet(_)
+        | Expr::Case { .. } => Ok(VisitRecursion::Continue),
+
+        Expr::ScalarFunction(scalar_function) => match &scalar_function.func_def {
+            ScalarFunctionDefinition::BuiltIn(fun) => match fun.volatility() {
+                Volatility::Immutable => Ok(VisitRecursion::Continue),
+                Volatility::Stable | Volatility::Volatile => {
+                    is_applicable = false;
+                    Ok(VisitRecursion::Stop)
+                }
+            },
+            ScalarFunctionDefinition::UDF(fun) => match fun.signature().volatility {
+                Volatility::Immutable => Ok(VisitRecursion::Continue),
+                Volatility::Stable | Volatility::Volatile => {
+                    is_applicable = false;
+                    Ok(VisitRecursion::Stop)
+                }
+            },
+            ScalarFunctionDefinition::Name(_) => {
+                internal_err!("Function `Expr` with name should be resolved.")
+            }
+        },
+
+        Expr::AggregateFunction { .. }
+        | Expr::Sort { .. }
+        | Expr::WindowFunction { .. }
+        | Expr::Wildcard { .. }
+        | Expr::Unnest { .. }
+        | Expr::Placeholder(_) => {
+            is_applicable = false;
+            Ok(VisitRecursion::Stop)
+        }
+    })
+    .unwrap();
+    is_applicable
+}
+
+// Below is all the logic necessary (I think) to convert a PhysicalExpr into a ZarrChunkFilter.
+// The logic is mostly copied from datafusion, and is simplified here for the zarr use case.
+pub struct ZarrFilterCandidate {
+    expr: Arc<dyn PhysicalExpr>,
+    projection: Vec<usize>,
+}
+
+struct ZarrFilterCandidateBuilder<'a> {
+    expr: Arc<dyn PhysicalExpr>,
+    file_schema: &'a Schema,
+    required_column_indices: BTreeSet<usize>,
+}
+
+impl<'a> ZarrFilterCandidateBuilder<'a> {
+    pub fn new(expr: Arc<dyn PhysicalExpr>, file_schema: &'a Schema) -> Self {
+        Self {
+            expr,
+            file_schema,
+            required_column_indices: BTreeSet::default(),
+        }
+    }
+
+    pub fn build(mut self) -> DataFusionResult<Option<ZarrFilterCandidate>> {
```

Comment:
Since there aren't any builder methods on this, maybe `TryFrom`? Also not super clear to my why it's an `Option`?

Reply:
Good question. So I think we need a builder struct because the way the columns indices get "extracted" from the predicate is through the call to `rewrite`, which takes a mut reference to self, on the below line. That function requires the `TreeNodeRewriter` trait on its argument, and you can see that as `pre_visit` is called, the indices get progressively filled. To be honest, I didn't dig all the way down to how this works, I just followed the steps they follow for parquet since I didn't want to risk breaking something.

Regarding the `Option`, you're right that it's not clear from the code here, I believe it's like that because of the code here, https://github.com/datafusion-contrib/arrow-zarr/pull/16/files#diff-d61c0a121604c7680df3d272638903a3fc21fee9ac3381e34b5285c02b9deaf0R202-R213, specifically because the `else` statement returns `None`. Since the type of `candidates` is `Vec<ZarrFilterCandidate>`, I think the call to `collect` coerces options into the inner type (or skips the value if it's `None`)? And that means the type of `candidate` must be `Option<...>`, so that the `if` and `else` statements return types match. Again, I mostly followed the parquet implementation.

I know that just following someone else's code and replicating it somewhat naively is not the best excuse haha, but like I said I wanted to minimize the risk of messing things up here, since I'm not yet comfortable with the code base. Overall does this all make sense?

Reply:
Also, when we start handling hive style partitions, and the logic gets more complicated, we might need to return a Ok(None) from `build` in some situations, I'm following the parquet logic but also simplifying it a lot (for now), so that might lead to code that looks a bit weird, temporarily.

Reply:
Cool cool, yeah that all sounds good to me to get started with. If we notice some perf issues w/ the cloning + rewriting we can reassess later.

Reply:
Yeah, I haven't been paying too much attention to everything that could impact performance so far, I'm thinking I'll revisit later when we have something fully functional.